File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> lastIndex;
        int maxLength = 0, start = 0;

        for (int j = 0; j < s.size(); j++) {
            if (lastIndex.find(s[j]) != lastIndex.end() && lastIndex[s[j]] >= start) {
                start = lastIndex[s[j]] + 1;  // Move start pointer to avoid repeating characters
            }
            lastIndex[s[j]] = j;  // Update last seen index
            maxLength = max(maxLength, j - start + 1);
        }
        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        maxi=i=0
        brr=dict()
        for j in range(len(s)):
            if s[j] in brr and brr[s[j]]>=i:
                i=brr[s[j]]+1
            brr[s[j]]=j
            maxi=max(maxi,j-i+1)
        return maxi 